tests/test_initial_data_preparation.py

```diff
@@ -0,0 +1,276 @@
+import pytest
+import pandas as pd
+from utils.initial_data_preparation import (
+    mark_incomplete_urls,
+    get_base_repo_url,
+    extract_and_flag_domains,
+)
+
+
+def test_get_base_repo_url():
+    # Test data setup
+    data = pd.DataFrame(
+        {
+            "repourl": [
+                None,  # Test for None input
+                "https://github.com/getdnsapi/stubby",
+                "https://github.com/namecoin",
+                "https://github.com/osresearch/heads/issues/540",
+                "https://git.sr.ht/~dvn/boxen",
+                "https://redmine.replicant.us/projects/replicant/wiki/"
+                "Tasks_funding#Graphics-acceleration",
+                "https://github.com/tdf/odftoolkit.git",
+                "https://git.savannah.gnu.org/git/mes.git",
+                "https://git.taler.net/git/libtalerutil.git",
+                "https://git.taler.net/taler-ios.git",
+                "just-a-string",
+                "http:///a-bad-url.com",
+                "https://github.com/getdnsapi/stubby.git#readme",
+                "https://github.com/getdnsapi/stubby.git?branch=main",
+                "https://gitlab.com/technostructures/kazarma/kazarma",
+                "https://gitlab.torproject.org/tpo/network-heal",
+                "https://codeberg.org/interpeer",
+                "https://codeberg.org/openEngiadina/geopub",
+                "https://codeberg.org/librEDA",
+                "https://framagit.org/incommon.cc",
+                "https://hydrillabugs.koszko.org/projects/haketilo/repository",
+                "https://git.replicant.us/replicant",
+                "https://gerrit.osmocom.org/plugins/gitiles/ims",
+            ],
+            "duplicate_flag": [False] * 23,
+            "unsupported_url_scheme": [False] * 23,
+            "incomplete_url_flag": [False] * 23,
+        }
+    )
+    data.loc[10:11, "unsupported_url_scheme"] = True  # Invalid URL formats
```

Comment:
I realise this is effectively copied from the previous code, nonetheless how about adding a brief note to say that 10 and 11 align with the entries?: 
```
                "just-a-string",
                "http:///a-bad-url.com",
```

Reply:
This is done.